pageserver/src/tenant.rs

```diff
@@ -1604,25 +1605,37 @@ impl Tenant {
             )
         })?;
         let timelines = self.timelines.lock().unwrap();
-        if let Some(timeline) = timelines.get(&timeline_id) {
-            let mut offloaded_timelines = self.timelines_offloaded.lock().unwrap();
-            if offloaded_timelines.remove(&timeline_id).is_none() {
-                warn!("timeline already removed from offloaded timelines");
-            }
-            info!("timeline unoffloading complete");
-            Ok(Arc::clone(timeline))
-        } else {
+        let Some(timeline) = timelines.get(&timeline_id) else {
             warn!("timeline not available directly after attach");
-            Err(TimelineArchivalError::Other(anyhow::anyhow!(
+            return Err(TimelineArchivalError::Other(anyhow::anyhow!(
                 "timeline not available directly after attach"
```

Comment:
This branch is a bit confusing to me: we're un-offloading a timeline, so why do we expect it to be present in `self.timelines`?  I thought an offloaded tenant would be in self.timelines_offloaded but not self.timelines?

Reply:
The `load_remote_timeline` call above calls `timeline_init_and_sync` which inserts the timeline into the `self.timelines` map. We only have to remove it from `timelines_offloaded`.

As there is no lock held between the call to `load_remote_timeline` above and our check below, there is the potential for race conditions, so I have not made it a panic.

Alternatively, one could also make `load_remote_timeline` and `timeline_init_and_sync` return the timeline. We'd probably immediately drop it in the other cases where the function is loaded, leading to some extra Arc refcount noise, but it should be negligible.

Reply:
ah, I see.  I had mis-remembered the scope of what load_remote_timeline does